core/store/src/trie/trie_storage.rs

```diff
@@ -307,7 +306,7 @@ pub trait TrieStorage {
 #[derive(Default)]
 pub struct TrieMemoryPartialStorage {
     pub(crate) recorded_storage: HashMap<CryptoHash, Arc<[u8]>>,
-    pub(crate) visited_nodes: RefCell<HashSet<CryptoHash>>,
+    pub(crate) visited_nodes: std::sync::RwLock<HashSet<CryptoHash>>,
```

Comment:
My gut feeling would have been `Mutex` to be a better match here, as it’s probably write-many-read-once. But I didn’t check all the uses and it’s likely unrelevant optimization anyway, so I’ll just leave that as an informational message if you want to think more about it.

Reply:
My reasoning here was that `RwLock` is a direct match to `RefCell` with regards to its API surface (`borrow` => `read`/`borrow_mut` => `write`.) We may indeed be able to revert this eventually; we'll see.

Reply:
SGTM :)